src/io/delimited.cr

```diff
@@ -0,0 +1,124 @@
+module IO
+  # An IO that wraps another IO, and only reads up to the beginning of a
+  # specified delimiter.
+  #
+  # This is useful for exposing part of an underlying stream to a client.
+  #
+  # ```
+  # io = MemoryIO.new "abc||123"
+  # delimited = IO::Delimited.new(io, read_delimiter: "||")
+  #
+  # delimited.gets_to_end # => "abc"
+  # delimited.gets_to_end # => nil
+  # io.gets_to_end        # => "123"
+  # ```
+  class Delimited
+    include IO
+
+    # If `sync_close` is true, closing this IO will close the underlying IO.
+    property? sync_close
+
+    getter read_delimiter
+    getter? closed : Bool
+
+    @delimiter_buffer : Bytes
+    @active_delimiter_buffer : Bytes?
```

Comment:
I think I understand the algorithm now, after writing it in paper :-)

So there's basically `@delimiter_buffer` that, when the first read byte matches with the first byte of the delimiter, gets filled. If it was filled to exactly the delimiter, we are done. Otherwise, set `@active_delimiter_buffer` to point to the first byte after the first one, and next reads will read from there. If we again find the first byte of the delimiter, we copy from `@active_delimiter_buffer` to the beginning of `@delimiter_buffer` and read what's remaining, compare again, etc. Right?

One thing we can try is to have `@active_delimiter_buffer` start as `Bytes.new(Pointer(UInt8).null, 0)` (maybe we could add a `Slice.null` method so that's easier to write) instead of `nil`. Then instead of checking if there's `@active_delimiter_buffer` we check if its size is greater than zero, and always move it forward. I think that could result in less comparisons, but I'm not sure. You could try making the change and benchmark it to see if it changes, and also check if there's less code in the end.

I actually did a benchmark with the change and it seems to be much faster, though I didn't run your specs. Here's the code: https://gist.github.com/asterite/9d87afd2bbe2749478079a56f3a85a06

Could you check that this works and it's faster?